src/Microsoft.AspNetCore.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -450,112 +449,116 @@ protected virtual string GenerateUniqueId()
             }
         }
 
-        private void RenderUnboundAttribute(string attributeName, Chunk attributeValueChunk)
+        private void RenderUnboundAttribute(TagHelperAttributeChunk attribute)
         {
             // Render children to provide IntelliSense at design time. No need for the execution context logic, it's
             // a runtime feature.
             if (_designTimeMode)
             {
-                if (attributeValueChunk != null)
+                if (attribute.Value != null)
                 {
-                    _bodyVisitor.Accept(attributeValueChunk);
+                    _bodyVisitor.Accept(attribute.Value);
                 }
 
                 return;
             }
 
-            // If we have a minimized attribute there is no value
```

Comment:
As discussed offline, a couple of these cases (minimized attributes and plain text attributes) fold into the preallocated case. That will allow considerable simplification of these blocks.